resharper/src/resharper-unity/CSharp/Daemon/Stages/Analysis/ComponentPropertyAccessProblemAnalyzer.cs

```diff
@@ -0,0 +1,385 @@
+using JetBrains.ReSharper.Feature.Services.Daemon;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Errors;
+using JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Dispatcher;
+using JetBrains.ReSharper.Plugins.Unity.Utils;
+using JetBrains.ReSharper.Psi;
+using JetBrains.ReSharper.Psi.CSharp.Tree;
+using JetBrains.ReSharper.Psi.CSharp.Util;
+using JetBrains.ReSharper.Psi.Resolve;
+using JetBrains.ReSharper.Psi.Tree;
+using JetBrains.Util;
+using QualifierEqualityComparer = JetBrains.ReSharper.Psi.CSharp.Impl.ControlFlow.ControlFlowWeakVariableInfo.QualifierEqualityComparer;
+
+namespace JetBrains.ReSharper.Plugins.Unity.CSharp.Daemon.Stages.Analysis
+{
+    [ElementProblemAnalyzer(
+        ElementTypes: new [] {typeof(IMethodDeclaration), typeof(ICSharpClosure)}, 
+        HighlightingTypes = new[] {typeof(InefficientPropertyAccessWarning)})]
+    public class ComponentPropertyAccessProblemAnalyzer : UnityElementProblemAnalyzer<ITreeNode>
+    {
+        private static readonly QualifierEqualityComparer ourComparer = new QualifierEqualityComparer();
+        
+        // NB : this analyzer invalidates all cached references (create a separated group of references which can be cached and
+        // pass them to quick fix) when encounter branches in control flow graph.
+        // e.g : invalidate before if, after else branch, after then branch, before loop, after loop section and so on.
+        
+        public ComponentPropertyAccessProblemAnalyzer([NotNull] UnityApi unityApi)
+            : base(unityApi)
+        {
+        }
+        
+        protected override void Analyze(ITreeNode node, ElementProblemAnalyzerData data, IHighlightingConsumer consumer)
+        {   
+            // container to collect groups of IReferenceExpressions which can be cached
+            var container = new PropertiesAccessContainer(consumer);
+            IBlock block = null;
+            
+            switch (node)
+            {
+                case IMethodDeclaration methodDeclaration:
+                    block = methodDeclaration.Body;
+                    break;
+                case ICSharpClosure closure:
+                    var body = closure.GetCodeBody();
+                    var exprBody = body.ExpressionBody;
+                    if (exprBody != null)
+                    {
+                        AnalyzeStatements(exprBody, container);
+                    }
+                    else
+                    {
+                        block = body.BlockBody;
+                    }
+                    break;
+            }
+            
+            if (block != null)
+            {
+                if (block.Descendants<IGotoStatement>().Any())
+                    return;
+                    
+                AnalyzeStatements(block, container);
+            }
+            
+            // cache all references which is not invalidated dut to no related expression was found
+            container.InvalidateCachedValues();
```

Comment:
I'd expect that `.InvalidateCachedValues` will clear some internal caches and definitely not that it will process some references and add highlightings. If it's not possible to compe up with a good name for its current functionality it'd probably be better to just split this method in two in order to reduce confusion over what it actually does.